### PR TITLE
SNOW-3031763: Add NUMERIC C to SQL conversions

### PR DESCRIPTION
**Pull request by sfc-gh-pbulawa**
### TL;DR

Expanded ODBC parameter binding to support cross-type conversions between C data types and SQL types, enabling conversions like integers to floats, floats to integers, and various numeric types to booleans.

### What changed?

Enhanced the ODBC driver's parameter binding system to handle a comprehensive set of type conversions:

**Boolean conversions**: Added support for converting all numeric C types (integers, floats, SQL_C_NUMERIC) to SQL_BIT, treating zero as false and non-zero as true.

**Numeric conversions**: Implemented bidirectional conversions between:
- Integer C types (SQL_C_SLONG, SQL_C_SBIGINT, etc.) to floating-point SQL types
- Floating-point C types (SQL_C_FLOAT, SQL_C_DOUBLE) to fixed-point SQL types with truncation
- SQL_C_NUMERIC to both fixed and floating-point SQL types with proper scale handling

**String conversions**: Extended SQL_C_BIT and SQL_C_NUMERIC support for VARCHAR binding.

**Helper functions**: Added `read_numeric_struct()` for parsing SQL_NUMERIC_STRUCT and `format_numeric_value()` for decimal string formatting.

**Error handling**: Added proper rejection of incompatible conversions (temporal/GUID types to numeric SQL types) with appropriate error codes.

### How to test?

Run the new E2E test suites:
- `c_integer_conversion_to_sql_fixed.cpp` - Integer to fixed-point conversions
- `c_float_conversion_to_sql_real.cpp` - Float to floating-point conversions  
- `c_numeric_types_conversion_to_sql_boolean.cpp` - Numeric to boolean conversions
- `c_temporal_incompatible_to_sql_*` - Verify proper rejection of invalid conversions

Test various scenarios including edge cases like negative values, zero, NULL indicators, and precision overflow.

### Why make this change?

This change improves ODBC compliance by implementing the full range of C-to-SQL data type conversions specified in the ODBC standard (Appendix D conversion tables). It enables applications to bind parameters using their natural C data types without requiring explicit type matching, making the driver more flexible and user-friendly. The changes also fix several behavior differences from the legacy driver, particularly around boolean conversions and numeric precision handling.

### CODE REVIEW

**Comment by sfc-gh-pbulawa** [comment]
* **#824** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/824?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#821** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/821?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#807** <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/807?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/snowflakedb/universal-driver/807?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: odbc/src/conversion/boolean.rs [hunk 1]
```diff
@@ -157,7 +160,72 @@ impl ReadODBC for SnowflakeBoolean {
         &self,
         binding: &'a ParameterBinding,
     ) -> Result<Self::Representation<'a>, JsonBindingError> {
-        Ok(read_unaligned::<u8>(binding) != 0)
+        let result = match binding.value_type {
+            CDataType::Bit => read_unaligned::<u8>(binding) != 0,
+            CDataType::TinyInt | CDataType::STinyInt => read_unaligned::<i8>(binding) != 0,
```
Comment (sfc-gh-makowalski) [inline odbc/src/conversion/boolean.rs:165]: `CDataType::Default` is not handled here — it falls through to the `_ =>` catch-all and returns `UnsupportedCDataType`. Before this change `read_odbc` had no match on `value_type` at all, so `Default` worked by accident. Now it would break for any app that binds `SQL_C_DEFAULT` to `SQL_BIT`.

`varchar.rs` already handles this: `CDataType::Default | CDataType::Char => ...`. Should be the same here:

```rust
CDataType::Default | CDataType::Bit => read_unaligned::<u8>(binding) != 0,
```

File: odbc_tests/common/include/conversion_checks.hpp [hunk 2]
```diff
@@ -233,4 +233,34 @@ inline void check_numeric_val_zero_from(const SQL_NUMERIC_STRUCT& numeric, int s
...
+    REQUIRE(!records.empty());
+    return;
+  }
+  REQUIRE(ret == SQL_SUCCESS);
+  ret = SQLExecute(stmt_handle.getHandle());
+  auto records = get_diag_rec(stmt_handle);
+  std::string sqlstate = records.empty() ? "(no diag)" : records[0].sqlState;
+  INFO("c_type=" << c_type << " sql_type=" << sql_type << " ret=" << ret << " sqlstate=" << sqlstate);
+  REQUIRE(ret == SQL_ERROR);
+  REQUIRE(!records.empty());
```
Comment (sfc-gh-makowalski) [inline odbc_tests/common/include/conversion_checks.hpp:261]: The other incompatible conversion helpers (`check_incompatible_conversion` here and `check_incompat<T>` in `timestamp_incompatible_checks.hpp`) both assert `CHECK(sqlstate == "07006")`. This one only checks that diag records exist but never verifies the SQLSTATE — a wrong error code (e.g. `HY000`) would pass silently.

Worth adding something like:
```cpp
CHECK(records[0].sqlState == "07006");
```
to both the `SQLBindParameter` and `SQLExecute` branches.

File: odbc_tests/tests/e2e/types/c_integer_conversion_to_sql_fixed.cpp [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+  REQUIRE_ODBC(ret, stmt);
+  SQLINTEGER val = 99999;
+  SQLLEN ind = 0;
+  ret = SQLBindParameter(stmt.getHandle(), 1, SQL_PARAM_INPUT, SQL_C_SLONG, SQL_INTEGER, 0, 0, &val, 0, &ind);
+  REQUIRE_ODBC(ret, stmt);
+  ret = SQLExecute(stmt.getHandle());
+
+  // Then the server rejects the value with an error
+  CHECK(ret == SQL_ERROR);
+}
```
Comment (sfc-gh-makowalski) [inline odbc_tests/tests/e2e/types/c_integer_conversion_to_sql_fixed.cpp:251]: This checks the call fails but not the SQLSTATE. The usual pattern in the other conversion tests is `CHECK(ret == SQL_ERROR)` + `CHECK(get_sqlstate(stmt) == "22003")` for numeric overflow. Same goes for the overflow tests in `c_float_conversion_to_sql_fixed.cpp` and `c_numeric_conversion_to_sql_fixed.cpp`.

**Review by sfc-gh-makowalski — CHANGES_REQUESTED** [review]
Nice work on the cross-type conversions — the Rust implementation is solid, especially the string-based `format_numeric_value` and the overflow-safe `read_numeric_struct`. The test coverage is thorough and follows our conventions well.

A few things to address before merging:

1. **`CDataType::Default` regression in `boolean.rs`** — the new match on `value_type` sends `Default` into the catch-all, breaking `SQL_C_DEFAULT` → `SQL_BIT` which worked before. Quick fix: `CDataType::Default | CDataType::Bit => ...`

2. **`check_incompatible_bindparam` missing SQLSTATE assertion** — the existing helpers (`check_incompatible_conversion`, `check_incompat<T>`) all verify `07006`. This new one should too.

3. **Overflow tests don't check SQLSTATE** — the `CHECK(ret == SQL_ERROR)` in the overflow rejection tests (`c_integer_conversion_to_sql_fixed`, `c_float_conversion_to_sql_fixed`, `c_numeric_conversion_to_sql_fixed`) should also check `get_sqlstate(stmt) == "22003"` to match the pattern used in other conversion tests.

4. **Char/WChar → boolean: parsed "NaN"/"inf" strings silently become `true`** — `parse::<f64>()` succeeds for these, and `NaN != 0.0` is true. The Float/Double branches already reject non-finite values with `is_finite()` — same check should apply after parsing in the Char/WChar branches. (Copilot flagged this too.)

File: odbc_tests/common/include/conversion_checks.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (sfc-gh-pbulawa) [reply odbc_tests/common/include/conversion_checks.hpp:261]: Done

Comment (sfc-gh-pbulawa) [reply odbc_tests/common/include/conversion_checks.hpp:261]: Done

File: odbc/src/conversion/boolean.rs (hunk 1 above)
Comment (sfc-gh-pbulawa) [reply odbc/src/conversion/boolean.rs:165]: Done

File: odbc/src/conversion/boolean.rs [hunk 4]
```diff
@@ -157,7 +160,72 @@ impl ReadODBC for SnowflakeBoolean {
...
+            CDataType::Numeric => {
+                let (mantissa, _scale) = read_numeric_struct(binding)?;
+                mantissa != 0
+            }
+            CDataType::Char => {
+                let s = read_char_str(binding)?;
+                let trimmed = s.trim();
+                if let Ok(v) = trimmed.parse::<f64>() {
+                    v != 0.0
+                } else {
```
[the remark this answers is not in the document]
Comment (sfc-gh-pbulawa) [reply odbc/src/conversion/boolean.rs:210]: Done

File: odbc/src/conversion/boolean.rs [hunk 5]
```diff
@@ -157,7 +160,72 @@ impl ReadODBC for SnowflakeBoolean {
...
+                    return UnsupportedCDataTypeSnafu {
+                        c_type: binding.value_type,
+                    }
+                    .fail();
+                }
+            }
+            CDataType::WChar => {
+                let s = read_wchar_str(binding)?;
+                let trimmed = s.trim();
+                if let Ok(v) = trimmed.parse::<f64>() {
```
[the remark this answers is not in the document]
Comment (sfc-gh-pbulawa) [reply odbc/src/conversion/boolean.rs:220]: Done

File: odbc_tests/tests/e2e/types/c_integer_conversion_to_sql_fixed.cpp (hunk 3 above)
Comment (sfc-gh-pbulawa) [reply odbc_tests/tests/e2e/types/c_integer_conversion_to_sql_fixed.cpp:251]: Done

File: odbc/src/conversion/real.rs [hunk 6]
```diff
@@ -354,6 +356,25 @@ impl ReadODBC for SnowflakeReal {
         let value = match binding.value_type {
             CDataType::Float => read_unaligned::<f32>(binding) as f64,
             CDataType::Double => read_unaligned::<f64>(binding),
+            CDataType::Long | CDataType::SLong => read_unaligned::<i32>(binding) as f64,
+            CDataType::Short | CDataType::SShort => read_unaligned::<i16>(binding) as f64,
+            CDataType::SBigInt => read_unaligned::<i64>(binding) as f64,
+            CDataType::ULong => read_unaligned::<u32>(binding) as f64,
+            CDataType::UShort => read_unaligned::<u16>(binding) as f64,
```
[the remark this answers is not in the document]
Comment (sfc-gh-pbulawa) [reply odbc/src/conversion/real.rs:365]: Done

**Review by sfc-gh-makowalski — APPROVED** [review]
